### PR TITLE
segment anything notebook

### PR DESCRIPTION
**Pull request by eaidova**

### CODE REVIEW

**Comment by eaidova** [comment]
<p>removed</p>

---

 <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1003/discussion/#comment-1504747578'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by eaidova** [comment]
<p>if we change that, we need to split decoders on 2 models (with mask and without it), at least we can suggest authors to remove flag as it really no need (mask multiplied on it, so zero * zero or zero * one still zero). Decoder part is lightweights, so there is no huge cost of this memory allocation</p>

---

 <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1003/discussion/#comment-1504747601'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by yury-gorbachev** [comment]
<p>I understand, just think that from model standpoint there is definitely opportunity for improvement, given that it is running in browser as well.</p><p> </p>

---

 <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1003/discussion/#comment-1504747601'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: README.md [hunk 1]
```diff
@@ -131,6 +131,7 @@ Demos that demonstrate inference on a particular model.
 |[234-encodec-audio-compression](notebooks/234-encodec-audio-compression/)<br>| Audio compression with EnCodec and OpenVINO™ | <img src=https://github.com/facebookresearch/encodec/raw/main/thumbnail.png width=225> |
 |[235-controlnet-stable-diffusion](notebooks/235-controlnet-stable-diffusion/)<br>| A Text-to-Image Generation with ControlNet Conditioning and OpenVINO™ | <img src=https://user-images.githubusercontent.com/29454499/224541412-9d13443e-0e42-43f2-8210-aa31820c5b44.png width=225> |
 |[236-stable-diffusion-v2](notebooks/236-stable-diffusion-v2/)<br>| Text-to-Image Generation and Infinite Zoom with Stable Diffusion v2 and OpenVINO™ | <img src=https://user-images.githubusercontent.com/29454499/228882108-25c1f65d-4c23-4e1d-8ba4-f6164280a3e3.gif width=225> |
+|[237-segment-anything](notebooks/237-segment-anything/)<br>| Prompt based object segmentation mask generation using Segment Anything and OpenVINO™ | <img src=https://user-images.githubusercontent.com/29454499/231465317-84e3de52-6d40-4cee-b6f2-ec7cbed1d72d.png width=225> |
```
Comment (eaidova) [inline README.md:134]:
```suggestion
|[237-segment-anything](notebooks/237-segment-anything/)<br>| Prompt based object segmentation mask generation using Segment Anything and OpenVINO™ | <img src=https://user-images.githubusercontent.com/29454499/231468849-1cd11e68-21e2-44ed-8088-b792ef50c32d.png width=225> |
```

File: notebooks/237-segment-anything/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+The notebook demonstrates how to work with model in 2 modes:
+
+* Interactive segmentation mode: in this demostration you can upload image and specify point related to object using [Gradio](https://gradio.app/) interface and as the result you get segmentation mask for specified point.
+The following image shows an example of the input text and the corresponding predicted image.
+![demo](https://user-images.githubusercontent.com/29454499/231464914-bd2a683c-28b2-44d4-960e-dce3e3ddebc3.png)
+
+* Automatic segmentation mode:  masks for the entire image can be generated by sampling a large number of prompts over an image.
+
+![demo2](https://user-images.githubusercontent.com/29454499/231465317-84e3de52-6d40-4cee-b6f2-ec7cbed1d72d.png)
```
Comment (eaidova) [inline notebooks/237-segment-anything/README.md:21]:
```suggestion
![demo2](https://user-images.githubusercontent.com/29454499/231468849-1cd11e68-21e2-44ed-8088-b792ef50c32d.png)
```

**Comment by raymondlo84** [comment]
image

Oh... interesting, where is the misalignment coming from? Thanks for trying these!

**Comment by raymondlo84** [comment]
Why are we exporting to onnx everytime?
Any idea?

import warnings
import torch
from openvino.tools import mo
from openvino.runtime import serialize, Core

core = Core()

with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
    warnings.filterwarnings("ignore", category=UserWarning)

    torch.onnx.export(sam.image_encoder, torch.zeros(1,3,1024,1024), "sam_image_encoder.onnx")

**Comment by OpenVINO-dev-contest** [comment]
> image Also, it seems we cannot run this remotely on a server.

you have to switch it to server's ip

**Comment by eaidova** [comment]
<p>it is not just decoder, it is decoder + prompt_encoder, that is why I call it generally predictor :)</p>

---

 <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1003/discussion/#comment-1505559827'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by eaidova** [comment]
<p>initially, I want to make it work standalone and remove any torch dependency (I think that our advantages to run it without torch dependency if model already converted),  but for automatic mask generation it is too hard and required too many code... It is also can serve like explanation how preprocessing work</p>

---

 <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1003/discussion/#comment-1505559846'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by eaidova** [comment]
> Why are we exporting to onnx everytime? Any idea? Shouldn't we just do it once?
> 
> ```
> import warnings
> import torch
> from openvino.tools import mo
> from openvino.runtime import serialize, Core
> 
> core = Core()
> 
> with warnings.catch_warnings():
>     warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
>     warnings.filterwarnings("ignore", category=UserWarning)
> 
>     torch.onnx.export(sam.image_encoder, torch.zeros(1,3,1024,1024), "sam_image_encoder.onnx")
> ```



> Why are we exporting to onnx everytime? Any idea? Shouldn't we just do it once?
> 
> ```
> import warnings
> import torch
> from openvino.tools import mo
> from openvino.runtime import serialize, Core
> 
> core = Core()
> 
> with warnings.catch_warnings():
>     warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
>     warnings.filterwarnings("ignore", category=UserWarning)
> 
>     torch.onnx.export(sam.image_encoder, torch.zeros(1,3,1024,1024), "sam_image_encoder.onnx")
> ```

good question, thanks for pointing that. I will add caching in next PR
